Add post-release blog article about MVP (#41723)
kubernetes/website: content/en/blog/_posts/2023-07-24-mixed-version-proxy-alpha.md

---
layout: blog

> title: "Kubernetes 1.28: Upgrade Kubernetes more safely by enabling Mixed Version Proxy (Alpha)"

Reviewer: How about:
```suggestion
title: "Kubernetes 1.28: A New (alpha) Mechanism For Safer API Upgrades"
```

Does that work? It's only a suggestion.

Author: Should we include the "Mixed Version Proxy" also in the title, like: `Kubernetes 1.28:  Mixed Version Proxy - A New (alpha) Mechanism For Safer API Upgrades` or is that too long? 

Also should we use `API upgrades` or `cluster upgrades` here?

Reviewer: You can pick. But you don't need Mixed Version Proxy in the title.

Author: Updated to use `Kubernetes 1.28: A New (alpha) Mechanism For Safer Cluster Upgrades`

date: 2023-05-24
slug: kubernetes-1-28-feature-mixed-version-proxy-alpha
---

**Author:** Richa Banker (Google)

> This blog describes “Mixed Version Proxy”, a new alpha feature in Kubernetes 1.28. Mixed Version Proxy enables a resource request to be served by the correct apiserver in cases where there are multiple apiservers at varied versions in a cluster, ex: during an upgrade of a cluster, when a runtime config changes and a rollout occurs.

Reviewer: How about:
```suggestion
This blog describes the _mixed version proxy_, a new alpha feature in Kubernetes 1.28. The
mixed version proxy enables an HTTP request for a resource to be served by the correct API server
in cases where there are multiple API servers at varied versions in a cluster. For example,
this is useful during a cluster upgrade, or when you're rolling out the runtime configuration of
the cluster's control plane.
```

I like the conciseness. You are very welcome to use something else that is concise but expresses things differently.

## What problem does this solve?
When a cluster undergoes an upgrade, the kube-apiservers existing at different versions in that scenario can serve different sets (groups, versions, resources) of built-in resources. A resource request made in this scenario may be served by any of the available apiservers, potentially resulting in the request ending up at an apiserver that may not be aware of the requested resource; consequently it being served a 404 not found error which is incorrect. Furthermore, incorrect serving of the 404 errors can lead to serious consequences such as namespace deletion being blocked incorrectly or objects being garbage collected mistakenly.

## How do we solve the problem?

> ![Alt text](https://github.com/kubernetes/website/assets/26771552/5c0fbb01-5e30-467d-81bf-5739be19b013?raw=true "MVP-flow-diagram")

Reviewer: - We recommend a figure shortcode.
- Please link to the rendered SVG asset so that Netlify CDN can serve it

Author: Added a figure shortcode.

> The new feature “Mixed Version Proxy” provides the kube-apiserver with the capability to proxy a request to a peer kube-apiserver which is aware of the requested resource and hence can serve the request. To do this, a new filter is added to the handler chain in the aggregator.

Reviewer: ```suggestion
The new feature “Mixed Version Proxy” provides the kube-apiserver with the capability to proxy a request to a peer kube-apiserver which is aware of the requested resource and hence can serve the request. To do this, a new filter has been added to the handler chain in the API server's aggregation layer.
```

2. To prevent indefinite rerouting of the request, a new header `X-Kubernetes-APIServer-Rerouted:true` is added to the original request once it is determined that the request cannot be served by the original apiserver that received the request and should therefore be proxied. The destination peer apiserver ensures to serve the request locally if it sees this header.
3. To set the network location of a kube-apiserver that peers will use to proxy requests, the value passed in `--advertise-address` or (when `--advertise-address` is unspecified) the `--bind-address` flag is used. For users with network configurations that would not allow communication between peer kube-apiservers using the addresses specified in these flags, there is an option to pass in the correct peer address as `--peer-advertise-ip` and `--peer-advertise-port` flags that are introduced in this feature.

## How do I enable this feature?
Following are the required steps to enable the feature:

> * Download the latest Kubernetes project (version [v1.28.0](http://v1.28.0/) or later)  

Reviewer: Doesn't look right. We can link to https://kubernetes.io/releases/download/ BTW

Author: Basically I want to emphasize that this feature is available on Kubernetes v1.28, what's a good way to communicate that?

Reviewer 2: I think the only ask here was to not use the `http://v1.28.0/` link. Since it looks like an invalid link accidentally created from the version number.

Author: Cool, updated to use https://kubernetes.io/releases/download/

* Pass the CA bundle that will be used by source kube-apiserver to authenticate destination kube-apiserver's serving certs using the flag `--peer-ca-file` on the kube-apiservers. Note: this is a required flag for this feature to work. There is no default value enabled for this flag.
* Pass the correct ip and port of the local kube-apiserver that will be used by peers to connect to this kube-apiserver while proxying a request. Use the flags `--peer-advertise-ip` and `peer-advertise-port` to the kube-apiservers upon startup. If unset, the value passed to either `--advertise-address` or `--bind-address` is used. If those too, are unset, the host's default interface will be used.

## What’s missing?
Currently we only proxy resource requests to a peer kube-apiserver when its determined to do so. Next we need to address how to work discovery requests in such scenarios. Right now we are planning to have the following capabilities for beta

* Merged discovery across all kube-apiservers
* Use an egress dialer for network connections made to peer kube-apiservers

## How can I learn more?

> - [KEP-4020: Mixed Version Proxy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy)

Reviewer: Even better: link to the documentation first, and the KEP second.

Author: Done, but the link will be broken since the documentation is not in the main branch yet?

Reviewer: That's not a problem - we never publish post-release blog articles before the release.

## How can I get involved?

> Reach us on slack [#sig-api-machinery](https://kubernetes.slack.com/messages/sig-api-machinery), or through the [mailing list](https://groups.google.com/forum/#!forum/kubernetes-sig-api-machinery). 

Reviewer: Slack (capital S); also please mention https://slack.k8s.io/ so that people can work out how to sign up.

Author: Done
